lib/logstash/filters/memcached.rb

```diff
@@ -67,28 +68,59 @@ class LogStash::Filters::Memcached < LogStash::Filters::Base
   # NOTE: in Memcached, a value of 0 (default) means "never expire"
   config :ttl, :validate => :number, :default => 0
 
+  # Tags the event on failure. This can be used in later analysis.
+  config :tag_on_failure, :validate => :string, :default => "_memcached_failure"
+
   public
 
   attr_reader :cache
 
   def register
-    if @ttl < 0
-      logger.error("ttl cannot be negative")
-      fail("invalid ttl: cannot be negative")
+    raise(LogStash::ConfigurationError, "'ttl' option cannot be negative") if @ttl < 0
+
+    @memcached_hosts = validate_connection_hosts
+    @memcached_options = validate_connection_options
+    begin
+      @cache = new_connection(@memcached_hosts, @memcached_options)
+    rescue => e
+      logger.error("failed to connect to memcached", hosts: @memcached_hosts, options: @memcached_options, message: e.message)
+      fail("failed to connect to memcached")
     end
-
-    @cache = establish_connection
-  end # def register
+    @connected = Concurrent::AtomicBoolean.new(true)
+    @connection_mutex = Mutex.new
+  end
 
   def filter(event)
-    set_success = do_set(event)
-    get_success = do_get(event)
+    unless valid_connection?
```

Comment:
Nitpick: the method name `valid_connection?` doesn't clue the reader into the side-effect of attempting to re-establish a broken connection. I tours around with suggesting this get broken up into multiple methods (called like `unless valid_connection? || reconnect`), but this too felt odd. Maybe a method name like `connection_available?` would at least hint that more work was being done, without changing your implementation.

Reply:
Agree!